File: backend/src/app/agent_hub/checkmk.py
"""Checkmk-agent output → snapshot enrichment (generic Linux nodes, §25/DR-10).

The linux agent pushes the raw output of the vendored ``check_mk_agent.linux``
as a ``checkmk_raw`` section (gzip+base64, see agent collect_checkmk). This
module parses the envelope plus a small, growing set of sections and rewrites
the push payload so the regular converters/checks see exactly the shapes the
FreeBSD collectors produce — downstream (SystemStatus, checks, metrics
history, UI) needs no linux branch, and evaluating a new Checkmk section is a
backend-only change (no agent rollout).

The parsers are own-built from the documented section formats. Do NOT copy
parse code from the GPLv2 Checkmk sources — we distribute container images,
and a derived work would drag the whole image under GPL (the vendored shell
script itself is attributed in THIRD-PARTY-NOTICES).

Everything here is pure and total: absent/broken data never raises, a section
that fails to parse simply leaves the original payload untouched (no data
must never alarm downstream).
"""

from __future__ import annotations

import base64
import binascii
import gzip
import re

# Decompressed cap. The agent caps raw output at 2 MB before gzip; anything
# larger here means a tampered/broken payload — drop it, don't parse it.
_MAX_RAW = 8_000_000

_HEADER_RE = re.compile(r"^<<<([a-zA-Z0-9_.-]+)((?::[a-zA-Z_]+\([^)]*\))*)>>>$")

# Cumulative (total, idle) jiffies from <<<kernel>>> — CPU% needs the delta
# between two pushes, so the hub keeps the previous tuple per instance.
CpuTicks = tuple[int, int]


def decode_raw(data: dict) -> str | None:
    """The decompressed checkmk_raw text of a push payload, or None."""
    section = data.get("checkmk_raw") or {}
    blob = section.get("output_gz_b64") if isinstance(section, dict) else None
    if not blob:
        return None
    try:
        raw = gzip.decompress(base64.b64decode(blob))
    except (ValueError, OSError, binascii.Error):
        return None
    if len(raw) > _MAX_RAW:
        return None
    return raw.decode("utf-8", "replace")


def parse_sections(text: str) -> dict[str, list[str]]:
    """Split Checkmk agent output into ``{section_name: lines}``.

    Piggyback blocks (``<<<<host>>>>`` … ``<<<<>>>>``) belong to other hosts
    and are skipped entirely. A repeated section name *concatenates* — the
    stock agent legitimately emits some sections twice (``lnx_if`` = ip-link
    block + counter/ethtool variant, ``df_v2`` = data + inode block), and the
    parsers are defensive enough that appended plugin output degrades to
    ignored lines rather than clobbering anything. Header options
    (``:sep(..)``, ``:cached(..)``) are dropped; the parsers below split
    whitespace-separated columns.
    """
    sections: dict[str, list[str]] = {}
    current: list[str] | None = None
    piggyback = False
    for line in text.splitlines():
        if line.startswith("<<<<") and line.endswith(">>>>"):
            piggyback = line != "<<<<>>>>"
            current = None
            continue
        if piggyback:
            continue
        m = _HEADER_RE.match(line)
        if m:
            current = sections.setdefault(m.group(1), [])
            continue
        if current is not None:
            current.append(line)
    return sections


def _parse_cpu(lines: list[str]) -> dict | None:
    """``<<<cpu>>>`` → the agent's loadavg shape.

    One line: ``load1 load5 load15 running/total lastpid [ncores]``.
    """
    if not lines:
        return None
    parts = lines[0].split()
    if len(parts) < 3:
        return None
    try:
        one, five, fifteen = (float(parts[i]) for i in range(3))
    except ValueError:
        return None
    cores = int(parts[5]) if len(parts) >= 6 and parts[5].isdigit() else 0
    return {"one": one, "five": five, "fifteen": fifteen, "cores": cores}


def _parse_kernel_ticks(lines: list[str]) -> CpuTicks | None:
    """Cumulative (total, idle) jiffies from the ``cpu`` summary row of
    ``<<<kernel>>>`` (/proc/stat format). iowait counts as idle — the CPU is
    not doing work while waiting on disk."""
    for line in lines:
        parts = line.split()
        if parts and parts[0] == "cpu" and len(parts) >= 5:
            try:
                vals = [int(x) for x in parts[1:]]
            except ValueError:
                return None
            idle = vals[3] + (vals[4] if len(vals) > 4 else 0)
            return (sum(vals), idle)
    return None


def cpu_pct_from_ticks(prev: CpuTicks | None, cur: CpuTicks | None) -> float | None:
    """Busy-CPU percent across two tick samples; None when not computable."""
    if prev is None or cur is None:
        return None
    d_total = cur[0] - prev[0]
    d_idle = cur[1] - prev[1]
    if d_total <= 0:
        return None
    pct = (d_total - d_idle) / d_total * 100.0
    return round(min(100.0, max(0.0, pct)), 1)


_MEM_RE = re.compile(r"^(\w+):\s+(\d+)")


def _parse_mem(lines: list[str]) -> dict | None:
    """``<<<mem>>>`` (/proc/meminfo, kB) → the agent's memory shape.

    Used = MemTotal - MemAvailable (the kernel's own reclaimable estimate);
    the Buffers+Cached fallback covers pre-3.14 kernels without MemAvailable.
    """
    kv: dict[str, int] = {}
    for line in lines:
        m = _MEM_RE.match(line)
        if m:
            kv[m.group(1)] = int(m.group(2))
    total_kb = kv.get("MemTotal", 0)
    if total_kb <= 0:
        return None
    avail_kb = kv.get("MemAvailable")
    if avail_kb is None:
        avail_kb = kv.get("MemFree", 0) + kv.get("Buffers", 0) + kv.get("Cached", 0)
    used_kb = max(0, total_kb - avail_kb)
    swap_total_kb = kv.get("SwapTotal", 0)
    swap_used_kb = max(0, swap_total_kb - kv.get("SwapFree", 0))
    total_mb = total_kb / 1024
    used_mb = used_kb / 1024
    swap_total_mb = swap_total_kb / 1024
    swap_used_mb = swap_used_kb / 1024
    return {
        "total_mb": round(total_mb, 1),
        "used_mb": round(used_mb, 1),
        "used_pct": round(used_mb / total_mb * 100, 1),
        "swap_total_mb": round(swap_total_mb, 1),
        "swap_used_mb": round(swap_used_mb, 1),
        "swap_used_pct": round(swap_used_mb / swap_total_mb * 100, 1) if swap_total_kb else 0.0,
    }


def _parse_df(lines: list[str]) -> list[dict] | None:
    """``<<<df_v2>>>`` (Checkmk ≥2.4; ``<<<df>>>`` before) → the agent's disks shape.

    Columns: device fstype size_kb used_kb avail_kb used% mountpoint. The
    section embeds an inode sub-block ([df_inodes_start]…[df_inodes_end]) —
    skipped. Only real block devices (/dev/…) or copy-on-write filesystems
    are kept: tmpfs/overlay/squashfs rows would flood a server's disk list.
    """
    rows: list[dict] = []
    in_inodes = False
    for line in lines:
        if line.startswith("[df_"):
            in_inodes = line.startswith("[df_inodes_start")
            continue
        if in_inodes:
            continue
        parts = line.split()
        if len(parts) < 7 or not parts[5].endswith("%"):
            continue
        device, fstype = parts[0], parts[1]
        if not device.startswith("/dev/") and fstype not in ("zfs", "btrfs", "xfs"):
            continue
        try:
            used_pct = float(parts[5].rstrip("%"))
        except ValueError:
            continue
        rows.append(
            {
                "device": device,
                # df escapes nothing — a mountpoint with spaces spans columns.
                "mountpoint": " ".join(parts[6:]),
                "used_pct": used_pct,
            }
        )
    return rows or None


# /proc/net/dev row inside <<<lnx_if>>>: "  eth0: 53756418 85747 0 5000 …"
# (16 counter columns; MAC/Speed lines from the ethtool blocks don't match
# because their values aren't digits-only).
_IF_COUNTER_RE = re.compile(r"^\s*([^\s:\[\]]+):\s*(\d[\d ]*)$")
_IF_IPLINK_RE = re.compile(r"^\d+:\s+([^:@\s]+)(?:@\S+)?:\s+<([^>]*)>")


def _parse_lnx_if(lines: list[str]) -> list[dict] | None:
    """``<<<lnx_if>>>`` (both variants concatenated) → the agent's interfaces
    shape.

    The ip-link block ([start_iplink]…[end_iplink]) provides flags/state and
    the primary IPv4; the counter rows (/proc/net/dev format) provide byte and
    error counters: rx bytes/packets/errs/… then tx bytes/packets/errs/…/colls.
    Loopback is dropped (noise on every host). Only interfaces with a counter
    row are reported — the counters drive the error-rate annotation downstream.
    """
    counters: dict[str, list[int]] = {}
    states: dict[str, str] = {}
    addrs: dict[str, str] = {}
    in_iplink = False
    current: str | None = None
    for line in lines:
        if line.startswith("[start_iplink]"):
            in_iplink = True
            continue
        if line.startswith("[end_iplink]"):
            in_iplink = False
            current = None
            continue
        if in_iplink:
            m = _IF_IPLINK_RE.match(line)
            if m:
                current = m.group(1)
                flags = m.group(2).split(",")
                states[current] = "up" if "UP" in flags else "down"
                continue
            am = re.match(r"\s+inet\s+([\d.]+)/", line)
            if am and current is not None:
                addrs.setdefault(current, am.group(1))
            continue
        m = _IF_COUNTER_RE.match(line)
        if m:
            fields = m.group(2).split()
            if len(fields) >= 14:
                counters[m.group(1)] = [int(x) for x in fields]
    rows = [
        {
            "name": name,
            "status": states.get(name, "up"),
            "address": addrs.get(name),
            "bytes_received": c[0],
            "bytes_transmitted": c[8],
            "in_errors": c[2],
            "out_errors": c[10],
            "collisions": c[13],
        }
        for name, c in counters.items()
        if name != "lo"
    ]
    return rows or None


def _parse_chrony(lines: list[str]) -> dict | None:
    """``<<<chrony>>>`` (chronyc tracking output) → the agent's ntp shape.

    "System time … fast/slow of NTP time" is the live offset (sign follows
    fast/slow); "RMS offset" stands in for jitter. An unsynchronised chrony
    reports stratum 0 / "Not synchronised" → synced False, and the ntp check's
    no-data sentinel (stratum < 0) stays reserved for "no chrony at all".
    """
    kv: dict[str, str] = {}
    for line in lines:
        if ":" in line:
            key, _, value = line.partition(":")
            kv[key.strip()] = value.strip()
    if "Stratum" not in kv:
        return None
    try:
        stratum = int(kv["Stratum"])
    except ValueError:
        return None
    offset_ms = 0.0
    m = re.match(r"([\d.]+)\s+seconds\s+(fast|slow)", kv.get("System time", ""))
    if m:
        offset_ms = round(float(m.group(1)) * 1000 * (1 if m.group(2) == "fast" else -1), 3)
    jitter_ms = 0.0
    jm = re.match(r"([\d.]+)\s+seconds", kv.get("RMS offset", ""))
    if jm:
        jitter_ms = round(float(jm.group(1)) * 1000, 3)
    peer = ""
    pm = re.search(r"\(([^)]+)\)", kv.get("Reference ID", ""))
    if pm:
        peer = pm.group(1)
    synced = stratum > 0 and "not synchronised" not in kv.get("Leap status", "").lower()
    return {
        "synced": synced,
        "stratum": stratum,
        "offset_ms": offset_ms,
        "jitter_ms": jitter_ms,
        "peer": peer,
    }


def _parse_systemd_units(lines: list[str]) -> list[dict] | None:
    """``<<<systemd_units>>>`` ([all] block) → the agent's services shape.

    Reported: running services (display + vital-service checks) and *failed*
    units (running False + failed marker → WARN check downstream). The
    hundreds of inactive/static units are deliberately dropped — an inactive
    oneshot is normal life, not a service list.
    """
    rows: list[dict] = []
    in_all = False
    for line in lines:
        if line.startswith("["):
            in_all = line.strip() == "[all]"
            continue
        if not in_all:
            continue
        parts = line.split(None, 4)
        if len(parts) < 4 or not parts[0].endswith(".service"):
            continue
        unit, _load, active, sub = parts[:4]
        name = unit[: -len(".service")]
        description = parts[4] if len(parts) > 4 else ""
        if active == "failed":
            rows.append(
                {"name": name, "description": description, "running": False, "failed": True}
            )
        elif active == "active" and sub == "running":
            rows.append({"name": name, "description": description, "running": True})
    return rows or None


def _parse_uptime(lines: list[str]) -> str | None:
    """``<<<uptime>>>`` (seconds since boot) → the human string the FreeBSD
    collector ships (uptime(1) style: "12 days, 3:07" / "3:07" / "42 mins")."""
    if not lines:
        return None
    try:
        secs = float(lines[0].split()[0])
    except (ValueError, IndexError):
        return None
    if secs < 0:
        return None
    mins = int(secs // 60)
    days, rem = divmod(mins, 1440)
    hours, minutes = divmod(rem, 60)
    if days:
        return f"{days} days, {hours}:{minutes:02d}"
    if hours:
        return f"{hours}:{minutes:02d}"
    return f"{minutes} mins"


def enrich_snapshot(
    data: dict, sections: dict[str, list[str]], prev_cpu_ticks: CpuTicks | None
) -> tuple[dict, CpuTicks | None]:
    """Rewrite a push payload with sections parsed from Checkmk output.

    Returns ``(new payload, current cpu ticks)`` — the input dict is never
    mutated. Only sections that parsed replace their orbit counterparts; on
    linux those hold zeros/junk (sysctl/pfctl don't exist there), so parsed
    Checkmk data wins. A parser returning None leaves the original key —
    partial Checkmk output degrades per-section, never wholesale.
    """
    out = dict(data)
    load = _parse_cpu(sections.get("cpu", []))
    if load is not None:
        out["loadavg"] = load
    mem = _parse_mem(sections.get("mem", []))
    if mem is not None:
        out["memory"] = mem
    # Live ubn1 (Checkmk 2.5.0p8) emits df_v2; older agent builds emit df.
    disks = _parse_df(sections.get("df_v2") or sections.get("df", []))
    if disks is not None:
        out["disks"] = disks
    uptime = _parse_uptime(sections.get("uptime", []))
    if uptime is not None:
        out["uptime"] = uptime
    interfaces = _parse_lnx_if(sections.get("lnx_if", []))
    if interfaces is not None:
        out["interfaces"] = interfaces
    ntp = _parse_chrony(sections.get("chrony", []))
    if ntp is not None:
        out["ntp"] = ntp
    services = _parse_systemd_units(sections.get("systemd_units", []))
    if services is not None:
        out["services"] = services
    ticks = _parse_kernel_ticks(sections.get("kernel", []))
    pct = cpu_pct_from_ticks(prev_cpu_ticks, ticks)
    if pct is not None:
        out["cpu"] = {"total_pct": pct}
    return out, ticks

File: orbit/lib/orbit/hub/checkmk.ex
defmodule Orbit.Hub.Checkmk do
  @moduledoc """
  Turn a Linux node's raw Checkmk-agent output into the same section shapes
  the firewall agents push (agent_hub/checkmk.py port, §25/DR-10).

  On a generic Linux server the orbit agent collects nothing itself — it runs
  the vendored upstream `check_mk_agent` and ships its stdout as gzip+base64
  under `checkmk_raw` (pure transport, so the GPLv2 script stays untouched).
  Until this module existed the hub's `Map.take` dropped that section on the
  floor without a word, while the agent's own zero-filled `cpu`/`memory`/
  `loadavg` sections were stored as-is: a Linux node enrolled, connected,
  reported hostname and uptime, and showed a permanent 0 % CPU / 0 % RAM —
  indistinguishable from an idle machine rather than from a broken import.

  Only the sections that map onto an existing Orbit shape are parsed —
  `mem`, `kernel`, `cpu`, `df_v2`, `lnx_if`, `uptime`. Everything else
  Checkmk emits (systemd units, plugin inventories, …) is ignored on
  purpose: a section Orbit has no home for would only invent checks nobody
  asked for.

  Note the split the Checkmk agent makes: `<<<cpu>>>` is `/proc/loadavg`
  plus the core count (so it feeds *loadavg*, not utilisation), while the
  `/proc/stat` tick counters live inside `<<<kernel>>>`. Reading utilisation
  out of `<<<cpu>>>` finds nothing at all.

  CPU needs two pushes: those ticks are cumulative, so utilisation is the
  delta against the previous sample. The previous sample is kept per
  instance in the hub cache entry itself (`checkmk_cpu`), which means a
  restart costs exactly one cycle of CPU data and nothing else.
  """

  require Logger

  @max_decompressed 8 * 1024 * 1024

  @doc """
  Decode + parse one `checkmk_raw` payload into `{sections, cpu_state}`.

  `prev_cpu` is the `checkmk_cpu` map from the instance's last push (or nil).
  Returns `{%{}, prev_cpu}` for anything unusable — a corrupt or oversized
  payload must leave the cached values alone rather than blank the box.
  """
  @spec parse(map() | nil, map() | nil) :: {map(), map() | nil}
  def parse(payload, prev_cpu \\ nil)

  def parse(%{"output_gz_b64" => b64}, prev_cpu) when is_binary(b64) do
    with {:ok, gz} <- decode64(b64),
         {:ok, text} <- gunzip(gz) do
      sections = split_sections(text)
      {cpu, cpu_state} = cpu_section(sections["kernel"], prev_cpu)

      parsed =
        %{
          "memory" => mem_section(sections["mem"]),
          "cpu" => cpu,
          "disks" => df_section(sections["df_v2"]),
          "interfaces" => lnx_if_section(sections["lnx_if"]),
          "uptime" => uptime_section(sections["uptime"]),
          "loadavg" => loadavg_section(sections["cpu"]),
          "ntp" => chrony_section(sections["chrony"]),
          "services" => systemd_section(sections["systemd_units"])
        }
        |> Enum.reject(fn {_k, v} -> v in [nil, [], %{}] end)
        |> Map.new()

      {parsed, cpu_state}
    else
      {:error, reason} ->
        Logger.warning("checkmk.parse_failed reason=#{reason}")
        {%{}, prev_cpu}
    end
  end

  def parse(_payload, prev_cpu), do: {%{}, prev_cpu}

  @doc """
  Just the decompressed agent output text (nil when unusable). Used to retain
  the raw dump for the instance's Checkmk view — separate from `parse/2` so
  its callers and tests are untouched (the extra gunzip is cheap).
  """
  @spec raw_text(map() | nil) :: String.t() | nil
  def raw_text(%{"output_gz_b64" => b64}) when is_binary(b64) do
    with {:ok, gz} <- decode64(b64),
         {:ok, text} <- gunzip(gz) do
      text
    else
      _ -> nil
    end
  end

  def raw_text(_), do: nil

  # -- decoding ---------------------------------------------------------------

  defp decode64(b64) do
    case Base.decode64(b64) do
      {:ok, bin} -> {:ok, bin}
      :error -> {:error, "bad base64"}
    end
  end

  # A zip bomb would otherwise be decompressed in full before we look at it.
  defp gunzip(gz) do
    text = :zlib.gunzip(gz)

    if byte_size(text) > @max_decompressed,
      do: {:error, "decompressed payload too large"},
      else: {:ok, text}
  rescue
    _ -> {:error, "bad gzip"}
  end

  @doc """
  Split Checkmk's `<<<section>>>` stream into `%{"name" => [lines]}`.

  A section header may carry options (`<<<lnx_if:sep(58)>>>`) and the same
  section can appear more than once (`df_v2` emits a second block for
  inodes) — options are dropped and repeats are concatenated, which matches
  how the upstream parsers read them.
  """
  def split_sections(text) when is_binary(text) do
    text
    |> String.split("\n")
    |> Enum.reduce({nil, %{}}, fn line, {current, acc} ->
      case section_header(line) do
        {:ok, name} -> {name, Map.put_new(acc, name, [])}
        :no -> {current, add_line(acc, current, line)}
      end
    end)
    |> elem(1)
    |> Map.new(fn {k, lines} -> {k, Enum.reverse(lines)} end)
  end

  defp section_header(line) do
    with true <- String.starts_with?(line, "<<<"),
         [name | _] <- line |> String.trim() |> String.trim_leading("<") |> String.split(">") do
      {:ok, name |> String.split(":", parts: 2) |> hd()}
    else
      _ -> :no
    end
  end

  defp add_line(acc, nil, _line), do: acc
  defp add_line(acc, section, line), do: Map.update(acc, section, [line], &[line | &1])

  # -- sections ---------------------------------------------------------------

  # <<<mem>>> is /proc/meminfo verbatim: "MemTotal:  1940712 kB".
  @doc false
  def mem_section(nil), do: nil

  def mem_section(lines) do
    values =
      for line <- lines,
          [key, rest] <- [String.split(line, ":", parts: 2)],
          {kb, _} <- [Integer.parse(String.trim(rest))],
          into: %{},
          do: {key, kb}

    total = values["MemTotal"]
    # MemAvailable is the honest "how much can a program still get" number —
    # MemFree alone counts cache as used and reports a healthy box at 95%.
    available = values["MemAvailable"] || values["MemFree"]

    if is_integer(total) and total > 0 and is_integer(available) do
      used_pct = (total - available) / total * 100.0
      swap_total = values["SwapTotal"] || 0
      swap_free = values["SwapFree"] || 0

      %{
        "total_mb" => round(total / 1024),
        "used_mb" => round((total - available) / 1024),
        "used_pct" => Float.round(used_pct, 1),
        "swap_total_mb" => round(swap_total / 1024),
        "swap_used_pct" => swap_used_pct(swap_total, swap_free)
      }
    end
  end

  defp swap_used_pct(total, free) when is_integer(total) and total > 0,
    do: Float.round((total - free) / total * 100.0, 1)

  defp swap_used_pct(_total, _free), do: 0.0

  # <<<cpu>>> line 1 = /proc/loadavg + core count, line 2 = jiffies total.
  # Utilisation is the delta of the busy jiffies against the previous push.
  @doc false
  def cpu_section(nil, prev), do: {nil, prev}

  def cpu_section(lines, prev) do
    case jiffies(lines) do
      nil ->
        {nil, prev}

      {busy, total} ->
        state = %{"busy" => busy, "total" => total}

        pct =
          case prev do
            %{"busy" => pb, "total" => pt} when is_number(pb) and is_number(pt) ->
              d_total = total - pt
              d_busy = busy - pb
              # A counter reset (reboot) or a repeated sample yields no usable
              # delta — report nothing rather than a spike or a fake 0%.
              if d_total > 0 and d_busy >= 0, do: Float.round(d_busy / d_total * 100.0, 1)

            _ ->
              nil
          end

        {if(pct, do: %{"total_pct" => pct}), state}
    end
  end

  # Inside <<<kernel>>>, /proc/stat's aggregate line keeps its label:
  # "cpu  user nice system idle iowait irq softirq steal …". The per-core
  # rows ("cpu0", "cpu1") are skipped — only the aggregate is wanted.
  defp jiffies(lines) do
    Enum.find_value(lines, fn line ->
      case String.split(line) do
        ["cpu" | rest] ->
          nums = for f <- rest, {n, ""} <- [Integer.parse(f)], do: n

          if length(nums) >= 5 do
            total = Enum.sum(nums)
            idle = Enum.at(nums, 3, 0) + Enum.at(nums, 4, 0)
            {total - idle, total}
          end

        _ ->
          nil
      end
    end)
  end

  # Line 1 of <<<cpu>>>: "0.08 0.08 0.09 1/156 507014 2" — the trailing field
  # is the core count.
  @doc false
  def loadavg_section(nil), do: nil

  def loadavg_section([first | _]) do
    case String.split(first) do
      [one, five, fifteen | rest] ->
        with {o, _} <- Float.parse(one),
             {f, _} <- Float.parse(five),
             {ft, _} <- Float.parse(fifteen) do
          %{"one" => o, "five" => f, "fifteen" => ft, "cores" => cores(rest)}
        else
          _ -> nil
        end

      _ ->
        nil
    end
  end

  def loadavg_section(_), do: nil

  defp cores(rest) do
    case rest |> List.last() |> to_string() |> Integer.parse() do
      {n, ""} when n > 0 -> n
      _ -> 0
    end
  end

  # <<<df_v2>>>: device fs total_kb used_kb avail_kb use% mountpoint.
  # The inode block ([df_inodes_start]…) and lsblk block are skipped.
  @doc false
  def df_section(nil), do: []

  def df_section(lines) do
    lines
    |> Enum.reduce({false, []}, fn line, {skipping, acc} ->
      cond do
        String.starts_with?(line, "[df_inodes_start]") or String.starts_with?(line, "[df_lsblk") ->
          {true, acc}

        String.starts_with?(line, "[df_inodes_end]") or
            String.starts_with?(line, "[df_lsblk_end]") ->
          {false, acc}

        skipping ->
          {true, acc}

        true ->
          {false, prepend_df(acc, line)}
      end
    end)
    |> elem(1)
    |> Enum.reverse()
    |> Enum.uniq_by(& &1["mountpoint"])
  end

  defp prepend_df(acc, line) do
    case String.split(line) do
      [device, _fs, total, used, _avail, pct, mount | _] ->
        with {total_kb, _} <- Integer.parse(total),
             {used_kb, _} <- Integer.parse(used),
             {used_pct, _} <- Integer.parse(String.trim_trailing(pct, "%")),
             true <- total_kb > 0 do
          [
            %{
              "device" => device,
              "mountpoint" => mount,
              "total_mb" => round(total_kb / 1024),
              "used_mb" => round(used_kb / 1024),
              "used_pct" => used_pct * 1.0
            }
            | acc
          ]
        else
          _ -> acc
        end

      _ ->
        acc
    end
  end

  # <<<lnx_if>>> carries an `ip link` dump plus a counter table
  # ("eth0: rx_bytes rx_packets … tx_bytes …"). Only the counter table maps
  # onto the interface shape the UI and the metric series already use.
  @doc false
  def lnx_if_section(nil), do: []

  def lnx_if_section(lines) do
    up = link_states(lines)

    for line <- lines,
        [name, rest] <- [String.split(line, ":", parts: 2)],
        name = String.trim(name),
        name != "",
        not String.contains?(name, " "),
        nums = for(f <- String.split(rest), {n, ""} <- [Integer.parse(f)], do: n),
        length(nums) >= 10 do
      %{
        "name" => name,
        "status" => if(Map.get(up, name, true), do: "up", else: "down"),
        "address" => nil,
        "bytes_received" => Enum.at(nums, 0, 0),
        "bytes_transmitted" => Enum.at(nums, 8, 0)
      }
    end
  end

  # "[eth0]" blocks end with "Link detected: yes|no".
  defp link_states(lines) do
    lines
    |> Enum.reduce({nil, %{}}, fn line, {current, acc} ->
      trimmed = String.trim(line)

      cond do
        String.starts_with?(trimmed, "[") and String.ends_with?(trimmed, "]") ->
          {trimmed |> String.trim_leading("[") |> String.trim_trailing("]"), acc}

        current && String.starts_with?(trimmed, "Link detected:") ->
          {current, Map.put(acc, current, String.contains?(trimmed, "yes"))}

        true ->
          {current, acc}
      end
    end)
    |> elem(1)
  end

  # <<<chrony>>> is `chronyc tracking` output. Mapped onto the same ntp shape
  # the FreeBSD boxes report, so ntp_check/1 works unchanged on Linux.
  @doc false
  def chrony_section(nil), do: nil

  def chrony_section(lines) do
    fields =
      for line <- lines,
          [k, v] <- [String.split(line, ":", parts: 2)],
          into: %{},
          do: {String.trim(k), String.trim(v)}

    with {stratum, _} <- Integer.parse(fields["Stratum"] || ""),
         true <- stratum > 0 do
      %{
        "stratum" => stratum,
        # "0.000057593 seconds fast of NTP time" → milliseconds.
        "offset_ms" => system_time_ms(fields["System time"]),
        "peer" => fields["Reference ID"],
        # Stratum 0 means "not synchronised" in chrony's own reporting; any
        # real stratum with a reference means the clock is disciplined.
        "synced" => true
      }
    else
      _ ->
        # chrony present but unsynchronised — WARN, never a missing check.
        if fields["Stratum"], do: %{"stratum" => 0, "synced" => false, "offset_ms" => 0.0}
    end
  end

  defp system_time_ms(nil), do: 0.0

  defp system_time_ms(text) do
    case Float.parse(String.trim(text)) do
      {secs, _} -> Float.round(secs * 1000, 3)
      :error -> 0.0
    end
  end

  # <<<systemd_units>>> [status] block: one "● unit.service" header per unit
  # followed by an indented "Active: <state> (...)" line. Only failed units
  # are reported — a service list of 400 units would drown the box's own
  # services view and emit hundreds of checks.
  @doc false
  def systemd_section(nil), do: []

  def systemd_section(lines) do
    lines
    |> Enum.reduce({nil, []}, fn line, {current, acc} ->
      trimmed = String.trim(line)

      cond do
        String.starts_with?(trimmed, "●") ->
          {trimmed |> String.trim_leading("●") |> String.trim() |> unit_name(), acc}

        current && String.starts_with?(trimmed, "Active:") ->
          failed? = String.contains?(trimmed, "failed")

          if failed?,
            do: {nil, [%{"name" => current, "description" => current, "running" => false} | acc]},
            else: {nil, acc}

        true ->
          {current, acc}
      end
    end)
    |> elem(1)
    |> Enum.reverse()
  end

  defp unit_name(header), do: header |> String.split(" ", parts: 2) |> hd()

  # <<<uptime>>>: "738381.32 1470462.76" — seconds since boot, idle seconds.
  #
  # Returns the SAME shape the firewall agent pushes: the human string
  # `uptime(1)` prints ("8 days, 12:33"), not a map. The detail page renders
  # this value directly, so a map crashed the whole Overview tab with a
  # Phoenix.HTML.Safe protocol error.
  @doc false
  def uptime_section(nil), do: nil

  def uptime_section([first | _]) do
    case first |> String.split() |> List.first() |> to_string() |> Float.parse() do
      {secs, _} when secs > 0 -> humanize_uptime(round(secs))
      _ -> nil
    end
  end

  def uptime_section(_), do: nil

  defp humanize_uptime(total) do
    days = div(total, 86_400)
    hours = total |> rem(86_400) |> div(3600)
    minutes = total |> rem(3600) |> div(60)
    clock = "#{hours}:#{String.pad_leading(to_string(minutes), 2, "0")}"

    case days do
      0 -> clock
      1 -> "1 day, #{clock}"
      n -> "#{n} days, #{clock}"
    end
  end
end
